src/passes/OptimizeInstructions.cpp

```diff
@@ -350,6 +350,46 @@ struct OptimizeInstructions
           return inner;
         }
       }
+      {
+        // x <<>> (C & (31 | 63))   ==>   x <<>> C'
+        // x <<>> (y & (31 | 63))   ==>   x <<>> y
+        // where '<<>>':
+        //   '<<', '>>', '>>>'. 'rotl' or 'rotr'
+        Type type = curr->type;
+
+        if (type.isInteger()) {
+          BinaryOp op;
+          Const* c;
+          Expression *x, *y;
+
+          if (matches(curr, binary(&op, any(&x), constant(&c))) &&
+              Abstract::hasAnyShift(type, op)) {
+            // truncate RHS constant
+            // x <<>> (C & (31 | 63))   ==>   x <<>> C'
+            c->value = c->value.and_(
+              Literal::makeFromInt32(type.getByteSize() * 8 - 1, type));
+            // x <<>> 0   ==>   x
+            if (c->value.isZero()) {
+              return x;
+            }
+          } else if (matches(
+                       curr,
+                       binary(&op,
+                              any(&x),
+                              binary(Abstract::And, any(&y), constant(&c)))) &&
+                     Abstract::hasAnyShift(type, op)) {
+            // x <<>> (y & (31 | 63))   ==>   x <<>> y
```

Comment:
It would be good to split these comments into two lines, one for the i32 case and one for the i64 case.